### PR TITLE
[GC] Add test for pause time record of ZGC class unloading

### PR DESCRIPTION
**Pull request by weixlu**
Summary: Add a test to check ZGC MXBean reports pause time of
class unloading

Reviewed-by: albert.th, mmyxym

Test Plan: gc/z/

Issue: https://github.com/alibaba/dragonwell11/pull/181

### CODE REVIEW

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    private static long reportedCycles = 0;
+    private static long reportedPauses = 0;
+
+    public static void main(String[] args) throws Exception {
+        unloadClassesFrequency = Long.parseLong(args[0]);
+
+        addListener();
+
+        // trigger gc
+        while (cycleId.get() < triggeredCycles) {
```
Comment (tanghaoth90) [inline test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java:61]: You may want to consider a simpler implementation to trigger ZGC by `wb.youngGC()` without using complex `addListener()`.
I suggest to remove `case GCCause::_wb_young_gc:` in `ZHeap::should_unload_class()`.

**Review by tanghaoth90 — CHANGES_REQUESTED** [review]
```
bool ZHeap::should_unload_class() {
  if (!ClassUnloading) return false;

  // ignore ZUnloadClassesFrequency if implied by the GC cause
  const GCCause::Cause cause = ZCollectedHeap::heap()->gc_cause();
  switch (cause) {
    case GCCause::_wb_young_gc:
    case GCCause::_wb_conc_mark:
    case GCCause::_wb_full_gc:
    case GCCause::_dcmd_gc_run:
    case GCCause::_java_lang_system_gc:
    case GCCause::_full_gc_alot:
    case GCCause::_scavenge_alot:
    case GCCause::_jvmti_force_gc:
    case GCCause::_metadata_GC_clear_soft_refs:
      return true;
  }

  return (ZUnloadClassesFrequency != 0) && ((ZGlobalSeqNum-1) % ZUnloadClassesFrequency == 0);
}
```

delete
```
    case GCCause::_wb_young_gc:
    case GCCause::_wb_conc_mark:
    case GCCause::_dcmd_gc_run:
    case GCCause::_scavenge_alot:
    case GCCause::_jvmti_force_gc:
```

**Review by mmyxym — APPROVED** [review]

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    public static void main(String[] args) throws Exception {
+        unloadClassesFrequency = Long.parseLong(args[0]);
+
+        // trigger gc
+        for (int i = 0; i < triggeredCycles; i++) {
+            wb.youngGC();
+        }
+
+        collectStatistics();
```
Comment (tanghaoth90) [inline test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java:64]: is it possible to move `collectStatistics();` and `verify();` into the `for` loop in order to verify each GC cycle?

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        long reportedCycles = 0;
+        long reportedPauses = 0;
+        for (final GarbageCollectorMXBean gc : ManagementFactory.getGarbageCollectorMXBeans()) {
+            if (gc.getName().equals("ZGC Cycles")) {
+                reportedCycles = gc.getCollectionCount();
+            } else if (gc.getName().equals("ZGC Pauses")) {
+                reportedPauses = gc.getCollectionCount();
+            }
+        }
+        if (reportedCycles == 0 || reportedPauses == 0) {
```
Comment (tanghaoth90) [inline test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java:74]: can be removed. we are not sure if `wb.youngGC();` can trigger GC in time

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+ * @requires vm.gc.Z & !vm.graal.enabled
+ * @summary Test whether ZGC garbage collector MXBean reports pause time and count of class unloading
+ * @library / /test/lib /runtime/testlibrary
+ * @build sun.hotspot.WhiteBox
+ * @run driver ClassFileInstaller sun.hotspot.WhiteBox
+ *                                sun.hotspot.WhiteBox$WhiteBoxPermission
+ * @run main/othervm -XX:+UseZGC -XX:+ClassUnloading -XX:ZUnloadClassesFrequency=2 -Xms128m -Xmx128m
+ *      -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xbootclasspath/a:.
+ *      gc.z.unloading.TestUnloadGarbageCollectorMXBean 2
+ * @run main/othervm -XX:+UseZGC -XX:+ClassUnloading -XX:ZUnloadClassesFrequency=13 -Xms128m -Xmx128m
```
Comment (tanghaoth90) [inline test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java:35]: add a new @run for `-XX:ZUnloadClassesFrequency=0`

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java (hunk 3 above)
Comment (tanghaoth90) [reply test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java:74]: you can change it into
```
if (reportedCycles == 0) {
    return reportedPauses == 0;
}
```

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            if (gc.getName().equals("ZGC Cycles")) {
+                reportedCycles = gc.getCollectionCount();
+            } else if (gc.getName().equals("ZGC Pauses")) {
+                reportedPauses = gc.getCollectionCount();
+            }
+        }
+        if (reportedCycles == 0 || reportedPauses == 0) {
+            return false;
+        }
+        if ((reportedPauses >= calculatePauses(reportedCycles - 1)) && (reportedPauses <= calculatePauses(reportedCycles))) {
```
Comment (tanghaoth90) [inline test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java:77]: return (reportedPauses >= calculatePauses(reportedCycles - 1)) && (reportedPauses <= calculatePauses(reportedCycles));

**Review by tanghaoth90 — APPROVED** [review]

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java (hunk 1 above)
Comment (weixlu) [reply test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java]: done

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java (hunk 2 above)
Comment (weixlu) [reply test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java]: done

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java (hunk 3 above)
Comment (weixlu) [reply test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java]: done

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java (hunk 4 above)
Comment (weixlu) [reply test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java:35]: done

File: test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java (hunk 5 above)
Comment (weixlu) [reply test/hotspot/jtreg/gc/z/unloading/TestUnloadGarbageCollectorMXBean.java]: done

**Pull request merged**